### ISSUE TITLE
hierarchical labels training / testing

### ISSUE THREAD

**Issue by wconnell** [KevinMusgrave/pytorch-metric-learning#68]
For `label_hierarchy_level` parameter use in `trainer` and `tester`, should passed dataset `__getitem__` return a 2D array, with column indices representing labels? 

I am doing this and passing `trainer(label_hierarchy_level=0)` and encountering a `RunTimeError` about mismatched tensor sizes (2) and (64) [batch size] in `get_all_triplets_indices(labels, ref_labels)`. Can't figure out if I am mispreprsenting data labels or if there is a bug.

**Comment by KevinMusgrave** [comment]
Something like this works for ```__getitem___```:

```python
def __getitem___(self, index):
    return img, np.array([label1, label2])
```

So it's actually a 1d array, but it becomes 2d when batched by the dataloader. Also, returning a list ```[label1, label2]``` doesn't work right now because the [label processing function](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/utils/common_functions.py#L64) is expecting a numpy array.

Note that you can also set ```label_hierarchy_level``` to ```"all"``` or to a list of integers which will be used to slice the columns of each batch of labels. However, none of the existing trainers are compatible with these settings (i.e. you'd have to write a custom trainer that has a method for calculating the loss based on multiple labels etc.)

**Comment by zawmintha** [comment]
> For `label_hierarchy_level` parameter use in `trainer` and `tester`, should passed dataset `__getitem__` return a 2D array, with column indices representing labels?
> 
> I am doing this and passing `trainer(label_hierarchy_level=0)` and encountering a `RunTimeError` about mismatched tensor sizes (2) and (64) [batch size] in `get_all_triplets_indices(labels, ref_labels)`. Can't figure out if I am mispreprsenting data labels or if there is a bug.

**Comment by vafatar** [comment]
I am creating an honest regular payment from home $1200-$2500/week , that is wonderful, beneath a year past i used to be idle in a very frightful economy. I give thanks God daily i used to be endowed these directions and currently it's my duty to pay it forward and share it with everybody, Here.............  **www.hitprofit1.com**

**Comment by beeandme** [comment]
Good , I hope I will start soon. my computer is a problem 
minimum what level of featured  computer needed ?
I also mix up bugging debugging sand stalk?
tanks , best of experience to build up for everyone.

**Comment by David711nava** [comment]
I have to much lagg on my fone

**Comment by Connect17** [comment]
I'm trying to come up with a first repository you friends have any Ideal what I could do? thanks!  Had to ask the first comment section I came too 👍

**Comment by Connect17** [comment]
now would I input this in cmd or where would we use the codes that come from GitHub? sorry kinda new

**Comment by wconnell** [comment]
If labels are hierarchical, `GlobalEmbeddingSpaceTester` does the same as `WithSameParentLabelTester`? Is there redundancy here?

**Comment by Connect17** [comment]
No There’s Not Redundancy I Don’t Guess. But How Would I Know I’m New..
And What Do You Mean By This? GlobalEmbeddingSpaceTester does the same as WithSameParentLabelTester?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by KevinMusgrave** [comment]
These fucking spam bots are driving me crazy!

@wconnell WithSameParentLabelTester is supposed to do something different. [The example in the documentation](https://kevinmusgrave.github.io/pytorch-metric-learning/testers/#withsameparentlabeltester) basically explains it. 

As a more specific example, consider a dataset with 1000 samples divided as follows:

- 400 dogs (animal)
- 300 cats (animal)
- 200 cars (vehicle)
- 100 trucks (vehicle)

GlobalEmbeddingSpaceTester:

- The "level 0" accuracy for a "dog" datapoint will be computed by considering all 999 other samples, using the {dog, cat, cars, trucks} labels.
- The "level 1" accuracy will also consider all 999 other samples, but using the {animal, vehicle} labels.

WithSameParentLabelTester:
- The "level 0" accuracy for a "dog" datapoint will be computed by considering only the 699 other samples that are "animals".
- There is no "level 1" accuracy, because level 1 has no parent labels (i.e. there's no level 2)

I haven't used WithSameParentLabelTester in a long time, and there isn't a test for it at the moment, so I can't guarantee that it works.

**Comment by wconnell** [comment]
Let's close this and duplicate in a new issue, I dont know why this is getting spammed.

I see, each sample does not require a hierarchical label in `WithSameParentLabelTester`.

Ok I'll follow up later, I think it's broken.

**Comment by KevinMusgrave** [comment]
@wconnell In v0.9.85.dev2 I fixed the bugs I mentioned above. I also fixed some other label hierarchy bugs, and WithSameParentLabelTester should at least run now.
